File: java/src/main/java/com/factset/protobuf/stach/extensions/RowStachExtensionBuilder.java
package com.factset.protobuf.stach.extensions;

import com.factset.protobuf.stach.v2.RowOrganizedProto;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.google.protobuf.InvalidProtocolBufferException;


public interface RowStachExtensionBuilder {

    /**
     * Sets the RowOrganizedPackage object.
     * @param pkg package object of type RowOrganizedPackage
     * @return builder instance
     */
    RowStachExtensionBuilder setPackage(RowOrganizedProto.RowOrganizedPackage pkg);

    /**
     * Sets the RowOrganizedPackage object by parsing raw object input.
     * @param pkg package object
     * @return builder instance
     */
    RowStachExtensionBuilder setPackage(Object pkg) throws JsonProcessingException, InvalidProtocolBufferException;

    /**
     * Set the RowOrganizedPackage object by parsing the input in string format.
     * @param pkgString string form of package object
     * @return builder instance
     */
    RowStachExtensionBuilder setPackage(String pkgString) throws InvalidProtocolBufferException;

    /**
     * Add the Row Organized Table to the package
     * @param tableId id of the table
     * @param table Row Organized Table object
     * @return Builder instance
     */
    RowStachExtensionBuilder addTable(String tableId, RowOrganizedProto.RowOrganizedPackage.Table table);

    /**
     * Add the Row Organized Table to the package
     * @param tableId id of the table
     * @param tableString Row Organized Table in string format
     * @return Builder instance
     */
    RowStachExtensionBuilder addTable(String tableId, String tableString) throws InvalidProtocolBufferException;

    /**
     * Add the Row Organized Table to the package
     * @param tableId id of the table
     * @param tableObject Row organized table object
     * @return Builder instance
     */
    RowStachExtensionBuilder addTable(String tableId, Object tableObject) throws JsonProcessingException, InvalidProtocolBufferException;

    /**
     * Builds the stach extension and returns the instance.
     * @return instance of row organized stach extension class.
     */
    StachExtensions build();
}

File: python/src/fds/protobuf/stach/extensions/v2/RowStachExtensionBuilder.py
import json

import google.protobuf.json_format
from fds.protobuf.stach.v2.RowOrganized_pb2 import RowOrganizedPackage

from fds.protobuf.stach.extensions.v2.RowOrganizedStachExtension import RowOrganizedStachExtension


class RowStachExtensionBuilder:

    def set_package(self, pkg):
        """
        Sets the package to the stach extension builder. The input package can be instance of Package, or
        string or object representation of the Package.

        :param pkg: Package object in string or object form.
        :return: builder instance
        """
        if (type(pkg) is RowOrganizedPackage):
            self.package = pkg
        elif (type(pkg) is str):
            package = RowOrganizedPackage()
            google.protobuf.json_format.Parse(pkg, package, ignore_unknown_fields=True, descriptor_pool=None)
            self.package = package
        else:
            pkg = json.dumps(pkg)
            package = RowOrganizedPackage()
            google.protobuf.json_format.Parse(pkg, package, ignore_unknown_fields=True, descriptor_pool=None)
            self.package = package

        return self

    def add_table(self, id, table):
        """
        Adds the table to the builder for conversion.

        :param id: id of the table.
        :param table: the inuput table object to be converted.
        :return: builder instance
        """
        if (type(table) is RowOrganizedPackage.Table):
            # we are checking if self.package is available, so that we can add the given table to the existing package.
            # if self.package is not available, we do create a package and add table to it.
            if (not hasattr(self, "package")):
                self.package = RowOrganizedPackage()
            empty_table = self.package.tables.get_or_create(id)
            empty_table.CopyFrom(table)
            return self
        elif (type(table) is str):
            pass
        else:
            table = json.dumps(table)

        tableInstance = RowOrganizedPackage.Table()
        google.protobuf.json_format.Parse(table, tableInstance, ignore_unknown_fields=True, descriptor_pool=None)

        # we are checking if self.package is available, so that we can add the given table to the existing package.
        # if self.package is not available, we do create a package and add table to it.
        if (not hasattr(self, "package")):
            self.package = RowOrganizedPackage()
        empty_table = self.package.tables.get_or_create(id)
        empty_table.CopyFrom(tableInstance)

        return self

    def get_package(self):
        """
        Returns the RowOrganizedPackage object set for the builder

        :return: RowOrganizedPackage instance
        """
        if (not hasattr(self, "package")):
            return None
        return self.package

    def build(self):
        """
        Builds and returns the instance of ColumnOrganizedStachExtension class

        """
        return RowOrganizedStachExtension(self.package)
